xds/internal/xdsclient/authority.go

```diff
@@ -304,8 +334,35 @@ func (a *authority) newConnectionError(err error) {
 	// For all resource types, for all resources within each resource type, and
 	// for all the watchers for every resource, propagate the connection error
 	// from the transport layer.
+	for _, rType := range a.resources {
+		for resourceName, state := range rType {
+			// If the connection/stream breaks after the resource has been requested,
+			// but before the response is received, we stop the timer here. When the
+			// stream is recreated, the transport layer will resend all previously
+			// requested resources, and we will start the timer in the `onResourceSend`
+			// callback.
+			if state.wState == watchStateRequested {
+				if ok := state.wTimer.Stop(); !ok {
+					a.logger.Warningf("wTimer.Stop() for resource %v returned false", resourceName)
```

Comment:
If we are unable to stop the timer here, it means that the timer fired and the expiry callback, `handleWatchTimerExpiry()`, was invoked. But since the expiry timer callback and this method grab the same lock to being with, it means that this method won the race for the lock. So, I think we should simply log the warning and not change the state here. It should continue to remain in `watchStateRequested`, so that once we release the lock here, the timer expiry callback would get the lock and would send the resource-not-found error to the watchers.

My suggestion would be to create another method (something like this and call it from here):

```golang
// stopWatchTimerLocked is invoked upon connection errors to stops watch timers for
// resources that have been requested, but not yet responded to by the management server.
//
// Caller must host a.mu.
func (a *authority) stopWatchTimerLocked() {
	for _, rType := range a.resources {
		for resourceName, state := range rType {
			if state.wState != watchStateRequested {
				return
			}
			if !state.wTimer.Stop() {
				// If the timer has already fired, it means that the timer watch expiry callback
				// is blocked on the same lock that we currently hold. Don't change the watch state
				// and instead let the watch expiry callback handle it.
				a.logger.Warningf("Watch timer for resource %v already fired. Ignoring here.", resourceName)
				continue
			}
			state.wState = watchStateStarted
		}
	}
}
```

What do you think?

Reply:
good catch. I had to add that extra check in the if condition there when it errored in a race test. I think this makes sense. Im renaming the method to be called `stopAllWatchTimersLocked()`